### DISCUSSION TITLE
NODE Control Adaptation

### DISCUSSION THREAD

**Discussion by johanndm1** [pnnl/neuromancer#208] [Q&A] [answered]
Hello,

I am trying to adapt your tutorial [Part 4: Learning neural ODE model and control policy for an unknown dynamical system.](https://colab.research.google.com/github/pnnl/neuromancer/blob/master/examples/control/Part_4_NODE_control.ipynb) to my own application. My code is essentially the same as that in the Jupyter notebook; the only differences being that I import my own data instead of generating it, and I had to use my own custom normalize/de-normalize functions. 

I am able to get good NODE system model identification:
![image](https://github.com/user-attachments/assets/f3acf94f-440b-4abd-9386-2bc01588e467)

However, my control policy is not as performant: 
![image](https://github.com/user-attachments/assets/9dd8e55a-d4fa-41de-b933-51c21d74ee64)

I am not sure what I am doing wrong and would appreciate any guidance. I've attached my notebook and data for your reference.
[NODE Control Adaptation.zip](https://github.com/user-attachments/files/17593937/NODE.Control.Adaptation.zip)

**Comment by drgona** [comment, marked as answer]
Hi @DJIDM. Your system ID looks spot on. Your control policy also does what it is supposed to do. The reason it is not tracking the reference perfectly is that the dynamics is slower than the change in the reference. I suggest increasing the period of change of the desired reference to match the system dynamics and you should see better tracking.

Also this is not an issue with the library. I suggest moving this to the [discussion channel](https://github.com/pnnl/neuromancer/discussions).

**Comment by johanndm1** [comment]
Hello Dr. Drgoňa,

Thank you for your response and feedback. 

Is there documentation on how to increase the period change of my desired reference to match my systems dynamics that I may refer to?

**Reply by drgona** [reply]
@DJIDM, you can change the sampling of your dataset; for instance, you can take only every 10th element of the time series, and this will give you a 10 times larger sampling interval.
